SYCL/Basic/reqd_work_group_size.cpp

```diff
@@ -12,6 +12,157 @@
+  });
+}
+
+template <TestKind Kind, int ReqdWGSizeX, int ReqdWGSizeY>
+void submitTest(queue &Q, range<2> GlobalRange) {
+  Q.submit([&](handler &CGH) {
+    CGH.parallel_for<ReqdWGSizeVariadic<Kind, ReqdWGSizeX, ReqdWGSizeY>>(
+        GlobalRange,
+        [=](item<2>)
+            [[sycl::reqd_work_group_size(ReqdWGSizeX, ReqdWGSizeY)]] {});
+  });
+}
+
+template <TestKind Kind, int ReqdWGSizeX>
+void submitTest(queue &Q, range<1> GlobalRange) {
+  Q.submit([&](handler &CGH) {
+    CGH.parallel_for<ReqdWGSizeVariadic<Kind, ReqdWGSizeX>>(
+        GlobalRange,
+        [=](item<1>) [[sycl::reqd_work_group_size(ReqdWGSizeX)]] {});
+  });
+}
+
+template <int... ReqdWGSizes>
+int runPositiveTest(queue &Q, range<sizeof...(ReqdWGSizes)> GlobalRange,
+                    range<sizeof...(ReqdWGSizes)> LocalRange) {
+  std::string TestCaseName =
+      "Test case ReqdWGSizeVariadic " +
+      GetRangeStr(range<sizeof...(ReqdWGSizes)>{ReqdWGSizes...}) + " & " +
+      GetRangeStr(LocalRange) + " (Positive with local range)";
+  try {
+    submitTest<TestKind::PositiveWithLocal, ReqdWGSizes...>(Q, GlobalRange,
+                                                            LocalRange);
+    Q.wait_and_throw();
+    return 0;
+  } catch (sycl::exception &E) {
+    std::cerr << TestCaseName
+              << " failed: unexpected sycl::exception: " << E.what()
+              << std::endl;
+    return 1;
+  } catch (...) {
+    std::cerr << TestCaseName << " failed: something unexpected has been caught"
+              << std::endl;
+    return 1;
+  }
+}
+
+template <int... ReqdWGSizes>
+int runPositiveTest(queue &Q, range<sizeof...(ReqdWGSizes)> GlobalRange) {
+  std::string TestCaseName =
+      "Test case ReqdWGSizeVariadic " +
+      GetRangeStr(range<sizeof...(ReqdWGSizes)>{ReqdWGSizes...}) +
+      " (Positive without local range)";
+  try {
+    submitTest<TestKind::PositiveWithoutLocal, ReqdWGSizes...>(Q, GlobalRange);
+    Q.wait_and_throw();
+    return 0;
+  } catch (sycl::exception &E) {
+    std::cerr << TestCaseName
+              << " failed: unexpected sycl::exception: " << E.what()
+              << std::endl;
+    return 1;
+  } catch (...) {
+    std::cerr << TestCaseName << " failed: something unexpected has been caught"
+              << std::endl;
+    return 1;
+  }
+}
+
+template <int... ReqdWGSizes>
+int runNegativeTest(queue &Q, range<sizeof...(ReqdWGSizes)> GlobalRange,
+                    range<sizeof...(ReqdWGSizes)> LocalRange) {
+  std::string TestCaseName =
+      "Test case ReqdWGSizeVariadic " +
+      GetRangeStr(range<sizeof...(ReqdWGSizes)>{ReqdWGSizes...}) + " & " +
+      GetRangeStr(LocalRange) + " (Negative)";
+  try {
+    submitTest<TestKind::Negative, ReqdWGSizes...>(Q, GlobalRange, LocalRange);
+    Q.wait_and_throw();
+    std::cerr << TestCaseName << " failed: no exception has been thrown\n";
+    return 1; // We shouldn't be here, exception is expected
+  } catch (nd_range_error &E) {
```

Comment:
Thank you for replacing those deprecated exceptions usage.
Shouldn't this one be replaced as well? It has such code in nd_range_error declaration:
"__SYCL2020_DEPRECATED("use sycl::exception with sycl::errc::nd_range instead.") nd_range_error"

Reply:
It should eventually be changed, but right now this is the exception we throw for this error case, so the test should be changed when we correct the behavior.